### DISCUSSION TITLE
Welcome to godot-go Discussions!

### DISCUSSION THREAD

**Discussion by pcting** [godot-go/godot-go#40] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by occamist** [comment]
@pcting is this repo still maintained? some links on readme is not working that are related to cgo references

**Reply by pcting** [reply]
not as active as i would like. i have an uncommitted branch  working against the godot 4 headers hoping this will yield better results.

**Reply by occamist** [reply]
cool, I am not a cgo expert but if the docs need to be maintained for Godot 4 version, I can help in order to learn the mysteries under the cgo. I have also raised this repository in gophers discord, I am hoping to get more people who are interested in Godot and Go, I saw godot-rust is more maintained so wanted to research why. It seems like most of the Go devs are missing non-GC-based language knowledge and inner workings of Go around cgo. I am hopeful that this repository can still be actively used for live projects maybe? 

are you using any codegen to generate bindings from Godot 4 headers? I see the rust community is using this without maintaining any C https://github.com/rust-lang/rust-bindgen

**Reply by pcting** [reply]
Thank you for the endorsement!

I've deprecated the Godot 3.x work and stuck it into the [godot-3 branch](https://github.com/godot-go/godot-go/tree/godot-3).

I've since replaced the master with the Godot 4 work.

Existing codegen has been all custom work through the use of https://github.com/alecthomas/participle/ to create a parser for parsing the C headers. This is what the `make generate` is doing. It generates C wrappers to be able to call C function pointers as well as generate classes and types expressed in the extension_api.json.

**Reply by occamist** [reply]
is there any difference between https://github.com/xlab/c-for-go and https://github.com/alecthomas/participle/ ? I am curious to observe it in terms of code generation

will check out the master and try it out. Xlab's C wrapper generator seems more mature, participle is saying it is experimental

**Reply by pcting** [reply]
they're very different projects. https://github.com/alecthomas/participle/ is used to generate parsers. you can use it to parse pretty much anything, but i have specified a grammer able to parse `gdnative_interface.h` here: https://github.com/godot-go/godot-go/blob/master/cmd/gdnativeparser/clang/parser.go

this converts the c header file into an [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree) where i can then generate custom wrapper code around.

here's an example of the parser for function pointers: https://github.com/godot-go/godot-go/blob/9d5a86402ae84d7cb3dfd41fadb18acd2ec6a8e5/cmd/gdextensionparser/clang/parser_test.go#L294-L308

i tried using c-for-go just a couple years back on my 3.x branch, but it just didn't play well with function pointers.

**Reply by pcting** [reply]
i'm generally go by trunk-based development and so i normally stick with main. i haven't had a chance to work on the project recently. i'm somewhat waiting for go to get [object pinning](https://github.com/golang/go/issues/46787) so that we can pass go memory around through cgo.

**Reply by pcting** [reply]
just wanted to update everyone that godot-go is now running against godot 4.2-dev6. the [Dodge the Creeps](https://github.com/godot-go/godot-go-demo-projects/tree/main/2d/dodge_the_creeps) demo has been updated and runs without crashes  on my linux machine.
